runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,482 @@
+local G = require('vim.lsp._snippet_grammar')
+local snippet_group = 'vim/snippet'
+local snippet_ns = vim.api.nvim_create_namespace(snippet_group)
+
+--- Returns the 0-based cursor position.
+---
+--- @return integer, integer
+local function cursor_pos()
+  local cursor = vim.api.nvim_win_get_cursor(0)
+  return cursor[1] - 1, cursor[2]
+end
+
+--- Resolves variables (like `$name` or `${name:default}`) as follows:
+--- - When a variable is unknown (i.e.: its name is not recognized in any of the cases below), return `nil`.
+--- - When a variable isn't set, return its default (if any) or an empty string.
+---
+--- Note that in some cases, the default is ignored since it's not clear how to distinguish an empty
+--- value from an unset value (e.g.: `TM_CURRENT_LINE`).
+---
+--- @param var string
+--- @param default string
+--- @return string?
+local function resolve_variable(var, default)
+  --- @param str string
+  --- @return string
+  local function expand_or_default(str)
+    local expansion = vim.fn.expand(str) --[[@as string]]
+    return expansion == '' and default or expansion
+  end
+
+  if var == 'TM_SELECTED_TEXT' then
+    -- Snippets are expanded in insert mode only, so there's no selection.
+    return default
+  elseif var == 'TM_CURRENT_LINE' then
+    return vim.api.nvim_get_current_line()
+  elseif var == 'TM_CURRENT_WORD' then
+    return expand_or_default('<cword>')
+  elseif var == 'TM_LINE_INDEX' then
+    return tostring(vim.fn.line('.') - 1)
+  elseif var == 'TM_LINE_NUMBER' then
+    return tostring(vim.fn.line('.'))
+  elseif var == 'TM_FILENAME' then
+    return expand_or_default('%:t')
+  elseif var == 'TM_FILENAME_BASE' then
+    -- Not using '%:t:r' since we want to remove all extensions.
+    local filename_base = expand_or_default('%:t'):gsub('%.[^%.]*$', '')
+    return filename_base
+  elseif var == 'TM_DIRECTORY' then
+    return expand_or_default('%:p:h:t')
+  elseif var == 'TM_FILEPATH' then
+    return expand_or_default('%:p')
+  end
+
+  -- Unknown variable.
+  return nil
+end
+
+--- Removes newlines from the given text.
+---
+--- @param text string|string[]
+--- @return string[]
+local function remove_newlines(text)
+  text = type(text) == 'string' and { text } or text
+  --- @cast text string[]
+  return vim.split(table.concat(text), '\n')
```

Comment:
Is removing newlines part of the spec? If a snippet contains them, shouldn't they be preserved?

---

Just as info: I suspect using `vim.tbl_filter` would be more efficient as it avoids the concat/split work. Fine to keep as well given that this is not run in an inner loop or performance sensitive.

Reply:
Maybe `remove_newlines` isn't the best name. I use this utility function to transform input like `foo\n\nbar` to `{ 'foo', '', 'bar' }`, since I rely on the number of snippet lines to compute the tabstop positions and to clean the input for `nvim_buf_set_text`.

Reply:
Maybe just `lines` then?

Reply:
I'll name it `text_to_lines`. `lines` feels to variable-y.